flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/WaitingForResources.java

```diff
@@ -144,12 +135,17 @@ public Logger getLogger() {
     }
 
     @Override
-    public void notifyNewResourcesAvailable() {
+    public void onNewResourcesAvailable() {
+        checkDesiredOrSufficientResourcesAvailable();
+    }
+
+    @Override
+    public void onNewResourceRequirements() {
         checkDesiredOrSufficientResourcesAvailable();
     }
 
     private void checkDesiredOrSufficientResourcesAvailable() {
-        if (context.hasDesiredResources(desiredResources)) {
+        if (context.hasDesiredResources()) {
```

Comment:
not directly related to this PR:
when `resourceStabilizationDeadline` is not null, should we skip scheduling `checkDesiredOrSufficientResourcesAvailable` (on line 162)?
Otherwise, we schedule as many checks as there are changes in resources.
Or am I missing something?

Reply:
yes we could refactor this. It probably won't result in wrong behavior, but still.

Instead of having the deadline we could instead schedule a single call to `checkDesiredOrSufficientResourcesAvailable` using the timeout as the delay.

Like this:
```
       if (context.hasSufficientResources()) {
            if (resourceStabilizationDeadline == null) {
                resourceStabilizationDeadline =
                        Deadline.fromNowWithClock(resourceStabilizationTimeout, clock);
                context.runIfState(
                        this,
                        this::checkDesiredOrSufficientResourcesAvailable,
                        resourceStabilizationDeadline.timeLeft());
            }
            if (resourceStabilizationDeadline.isOverdue()) {
                createExecutionGraphWithAvailableResources();
            }
        } else {
            // clear deadline due to insufficient resources
            resourceStabilizationDeadline = null;
        }
```

Anyway, @rkhachatryan you want to file a ticket for this?

Reply:
Done: https://issues.apache.org/jira/browse/FLINK-31601